src/utilities/Dealer.cpp

```diff
@@ -0,0 +1,122 @@
+    }
+    return result;
+}
+
+// Populate a matrix with scores
+std::vector<vector<double>> Dealer::getScoreMatrix(const std::vector<v::RobotView> &allRobots, const Dealer::FlagMap &flagMap) {
+    vector<vector<double>> scores;
+    scores.reserve(flagMap.size());
+    for (auto const& [roleName, dealerFlags] : flagMap) {
+        std::vector<double> row;
+        row.reserve(allRobots.size());
+        for (auto robot : allRobots) {
+            row.push_back(scoreForFlags(dealerFlags, robot));
+        }
+        scores.push_back(row);
+    }
+    return scores;
+}
+
+double Dealer::scoreForFlags(const std::vector<Dealer::DealerFlag> &dealerFlags, const v::RobotView &robot) {
+    double robotScore = 0;
+    for (auto flag : dealerFlags) {
+        robotScore += getScoreForFlag(robot, flag);
+    }
+    return robotScore;
+}
+
+double Dealer::getScoreForFlag(v::RobotView robot, Dealer::DealerFlag flag) {
+    double factor = getFactorForPriority(flag);
+    return factor * getDefaultFlagScores(robot, flag);
+}
+
+double Dealer::getFactorForPriority(const Dealer::DealerFlag &flag) {
+    switch (flag.priority) {
+        case DealerFlagPriority::LOW_PRIORITY: return 1.0;
+        case DealerFlagPriority::MEDIUM_PRIORITY: return 2.0;
+        case DealerFlagPriority::HIGH_PRIORITY: return 3.0;
+    }
+}
+
+// TODO 'invert' the matrix scores
+double Dealer::getDefaultFlagScores(const v::RobotView &robot, const Dealer::DealerFlag &flag) {
+    auto fieldWidth = field->get_field().get(FIELD_WIDTH);
+    auto fieldLength = field->get_field().get(FIELD_LENGTH);
+    switch (flag.title) {
+        case DealerFlagTitle::CLOSE_TO_THEIR_GOAL: 
+            return costForDistance(field->getDistanceToGoal(false, robot->getPos()), fieldWidth, fieldLength);
+        case DealerFlagTitle::CLOSE_TO_OUR_GOAL: 
+            return costForDistance(field->getDistanceToGoal(true, robot->getPos()), fieldWidth, fieldLength);
+        case DealerFlagTitle::CLOSE_TO_BALL: {
+            auto ball = world.getBall();
+            if (!ball) return 0.0;
+            return robot->getPos().dist(ball.value()->getPos());
+        }
+        case DealerFlagTitle::WITH_WORKING_BALL_SENSOR: return costForProperty(robot->isWorkingBallSensor());
+        case DealerFlagTitle::ROBOT_TYPE_50W: return costForProperty(robot->isFiftyWatt());
+        case DealerFlagTitle::ROBOT_TYPE_30W: return costForProperty(robot->isThirtyWatt());
+        case DealerFlagTitle::WITH_WORKING_DRIBBLER: return costForProperty(robot->isWorkingDribbler());
+        case DealerFlagTitle::READY_TO_INTERCEPT_GOAL_SHOT: {
+            // get distance to line between ball and goal
+            // TODO this method can be improved by choosing a better line for the interception.
+            LineSegment lineSegment = {world.getBall()->get()->getPos(), field->get_field().get(OUR_GOAL_CENTER)};
+            return control::ControlUtils::distanceToLine(robot->getPos(), lineSegment.start, lineSegment.end);
```

Comment:
return linesegment.distanceToLine(robot->getPos()) is better here. Try to avoid using controlUtils if possible.

Reply:
You can then also remove the include.

Reply:
We will probably want a call to some calculator that does line interception and trajectory planning magic in the  future though.

Reply:
ah I was looking for that function but I could not find it. might have overlooked it.